src/pocketmine/Player.php

```diff
@@ -3279,6 +3279,11 @@ public function sendWhisper(string $sender, string $message){
 	}
 
 	public function sendForm(Form $form) : void{
+		if($form->hasBeenSent()){
+			throw new \InvalidArgumentException("Cannot send the same form more than once, create a new one instead");
```

Comment:
But: Why? So i can not cache a Form and send it to a player? So it has to be recreated (rebuildt in the core) every time? Isn't that just a waste of calculation if the form contents would never change?

Reply:
The core code has no foolproof way to tell if your form is immutable (so to speak). You can clone the form object if you're concerned about the overhead.

Reply:
I have already explained in my previous comment:

> It is particularly tempting to hold a CustomForm instance and use the values there later, since the data cannot be easily extracted. I can foresee this is going to cause a lot of concurrency bugs that can be very difficult to identify. Therefore, it is suggested that, starting from the core, one of either approach ("resend same instance" or "new instance each resend") be adopted to avoid confusion.
>
> * Adopting the "resend same instance" approach should automatically call the clearResponseData() method after each onSubmit() call.
> * Adopting the "new instance each resend" approach should remove this method and add a flag to identify a Form as sent and throw an exception when attempting to use sendForm() with it.

As it has been noticed that resetting contents each time makes it very inconvenient to retain data (especially inconvenient for custom form data), the "new instance each resend" approach is considered more appropriate.

Reply:
Regarding the overhead of form creation overhead, perhaps you should compare it against encoding the form, sending it to the client, making the client display it and sending it back. As one wouldn't reasonably send forms very frequently, the impact on performance is negligible, so trying to cache forms for the sake of performance would be premature optimization, and therefore your comment is 4√(all evil), and so you are 8√(all evil).